engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReProcessingStateMachine.java

```diff
@@ -341,10 +323,12 @@ private TransactionOperation chooseOperationForEvent(
       final long position, final TypedRecord<?> currentEvent) {
     final TransactionOperation operationOnProcessing;
 
-    if (failedEventPositions.contains(position)) {
-      LOG.info(LOG_STMT_FAILED_ON_PROCESSING, currentEvent);
+    if (currentEvent.getValueType() == ValueType.ERROR) {
+      final var errorRecord = (ErrorRecord) currentEvent.getValue();
+      final var blacklistedInstance = errorRecord.getProcessInstanceKey();
+      LOG.info(LOG_STMT_FAILED_ON_PROCESSING, currentEvent, blacklistedInstance);
       operationOnProcessing =
-          () -> zeebeState.getBlackListState().tryToBlacklist(currentEvent, NOOP_LONG_CONSUMER);
+          () -> zeebeState.getBlackListState().blacklistProcessInstance(blacklistedInstance);
```

Comment:
This works :+1:

Alternatively, we could move the blacklisting to a regular event applier. Error events would be handled as any other event. We don't need to handle them differently in the stream processor.

What do you think?

Reply:
I just used that on the POC, but also thought about the applier but was not sure. Would be fine for me. Probably more elegant then this. I change it in the afternoon :)